src/creator.rs

```diff
@@ -97,53 +93,244 @@ impl<H: Hasher> Creator<H> {
+                    }
+                    continue;
+                }
+                _ = &mut delay => {
+                    break;
+                }
+            }
         }
-        // To create a new unit, we need to have at least >floor(2*N/3) parents available in previous round.
+
+        // To create a new unit, we need to have at least floor(2*N/3) + 1 parents available in previous round.
         // Additionally, our unit from previous round must be available.
-        let prev_round_index = (self.current_round - 1) as usize;
-        let threshold = (self.n_members * 2) / 3;
+        let threshold = (self.n_members * 2) / 3 + NodeCount(1);
 
-        self.n_candidates_by_round[prev_round_index] > threshold
-            && self.candidates_by_round[prev_round_index][self.node_ix].is_some()
+        while self.n_candidates_by_round[prev_round_index] < threshold || self.candidates_by_round[prev_round_index][self.node_ix].is_none() {
+             if let Some(u) = self.parents_rx.next().await {
+                self.add_unit(u.round(), u.creator(), u.hash());
+             } else {
+                warn!(target: "AlephBFT-creator", "{:?} get error as result from channel with parents.", self.node_ix);
+            }
+        }
     }
 
     pub(crate) async fn create(&mut self, mut exit: oneshot::Receiver<()>) {
-        let half_hour = Duration::from_secs(30 * 60);
-        let mut round: usize = 0;
-        let mut delay_fut = Delay::new((self.create_lag)(0)).fuse();
-        let mut delay_passed = false;
-        loop {
-            futures::select! {
-                unit = self.parents_rx.next() => {
-                    if let Some(u) = unit{
-                        self.add_unit(u.round(), u.creator(), u.hash());
+        for round in 0..self.max_round {
+            let mut ticker = time::interval(Duration::from_secs(30 * 60));
+            ticker.tick().await;
+            loop {
+                tokio::select! {
+                    _ = self.wait_until_ready(round) => {
+                        break;
                     }
-                },
-                _ = &mut delay_fut => {
-                    if delay_passed {
+                    _ = ticker.tick() => {
                         warn!(target: "AlephBFT-creator", "{:?} more than half hour has passed since we created the previous unit.", self.node_ix);
                     }
-                    delay_passed = true;
-                    delay_fut = Delay::new(half_hour).fuse();
+                    _ = &mut exit => {
+                        info!(target: "AlephBFT-creator", "{:?} received exit signal.", self.node_ix);
+                        return;
+                    }
+                }
+            }
+            self.create_unit(round);
+        }
+        warn!(target: "AlephBFT-creator", "{:?} Maximum round reached. Not creating another unit.", self.node_ix);
+    }
+}
+
+
+#[cfg(test)]
```

Comment:
If I read correctly there is no test here that actually would have failed before this change. Do you think you would be able to write such a test?

Reply:
Done